### PR TITLE
Improve ec2_info grain

### PR DESCRIPTION
**Pull request by TronPaul**
- unflatten the ec2 grains
- snake case all aws keys
- parse json values (for ami permissions)

### CODE REVIEW

**Comment by TronPaul** [comment]
Major changes to the grain output. I think the unflattened dictionary is easier to use than long keys. And it felt weird to me that only the additional info keys were snake cased. Interested in what other users of this grain think.

**Comment by mafrosis** [comment]
In theory I like the change. However changing grain names will break whatever states/modules/other external code is already using them.

Also, using the top-level `ec2` key - is it then possible to query a single grain directly? For example:

    host> salt-call grains.item ec2
    local:
      ec2: {u'availability_zone': u'ap-southeast-2b', 'public_ipv4': ..... }

**Comment by TronPaul** [comment]
@mafrosis what do you mean by single grain? You can do

    host> salt-call grains.get ec2
    local:
    ----------
    account_id:
        111111111111
    ami_id:
        ami-76e27e1e
    ami_launch_index:
        0
    ........

or if you want a specific underlying key

    host> salt-call grains.get ec2:availability_zone
    local:
        us-east-1c

**Comment by TronPaul** [comment]
I did just find an issue with this on Python 2.6.x. Dictionary comprehensions weren't added until 2.7. Need to find what versions of Python salt supports and test this code on the oldest version.

**Comment by cro** [comment]
Salt is supported under Python 2.6 and 2.7.

--cro

**Comment by TronPaul** [comment]
I can spin up an ec2 instance to test 2.6 < 2.6.9 if anyone thinks there's value in doing so, otherwise I'll assume a good run with 2.6.9 is sufficient.

**Comment by cro** [comment]
Mark,

We have lots of users on CentOS / RHEL 6, so if you are willing to check with the default Python on either of those distros that would be much appreciated.  It's 2.6.6 there I think.

**Comment by TronPaul** [comment]
Will do. I'll post an update when I have time to check it (should be later today). I plan on checking with Python 2.6.0.

Also, how do breaking changes get handled with salt-contrib? I'm curious as this would change the grain output and how to lookup ec2_info grains.

**Comment by cro** [comment]
> Will do. I'll post an update when I have time to check it (should be later today). I plan on checking with Python 2.6.0.
> 
> Also, how do breaking changes get handled with salt-contrib? I'm curious as this would change the grain output and how to lookup ec2_info grains.
> 

AFAIK you just need to document it clearly.

--cro

**Comment by TronPaul** [comment]
@cro, you've just made me realize that I made no changes to the documentation. I'll fix that up.

**Comment by cro** [comment]
> @cro <https://github.com/cro>, you've just made me realize that I made no changes to the documentation. I'll fix that up.
> 
Much appreciated!

--cro

File: grains/ec2_info.py [hunk 1]
```diff
@@ -28,24 +28,48 @@ def _call_aws(url):
...
+                try:
+                    data = json.loads(call_response)
+                    if isinstance(data, dict):
+                        data = _snake_caseify_dict(data)
+                    d[line] = data
+                except ValueError:
+                    d[line] = call_response
             else:
-                data["ec2_" + path.replace("/", "_")[:-1]] = line
+                return line
```
Comment (TronPaul) [inline grains/ec2_info.py]: Looking back on this I'm worried this isn't actually a base case. Will look into this.

**Comment by TronPaul** [comment]
Works on Python-2.6. Going to test that base case issue.

**Comment by TronPaul** [comment]
So the only thing that hits that `return line` is the public keys list at `public-keys/`. I'm trying to figure out how to refactor the code to make the recursive structure work.

**Comment by TronPaul** [comment]
Reading through the documentation for [Instance Metadata](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html), some of the paths don't follow the kind of recursive dict pattern. Makes a solution that works for everything difficult.

**Pull request merged**